### DISCUSSION TITLE
Consumption data upload

### DISCUSSION THREAD

**Discussion by bshaffer82** [SBFspot/SBFspot#464] [General]
I've created a consumption meter which can provide usage watts using a spark particle.  I've had SBFspot working for a couple of years now and I want to add the consumption data so it gets uploaded to PVOutupt.  

What do I need to do?
Can I just add the data into the consumption mysql table?  if so what's the required table format?  Do I need to tag the data with the same timestamp as the data coming out of Bluetooth data capture?

I'm an intermediate programmer but not that experienced with sqlite.

Thanks

**Comment by sbf-** [comment]
Yes, fill consumption table with Timestamp, Energy (Wh) and Power (W)
The only challenge is the upload when it's dark and there is no production. Upload stops when production stops, so this is still an issue. This requires changes to the code.

**Comment by bshaffer82** [comment]
What is the format for those three?  

Specifically, what is the date format?  I would rather push the data from python than try an recompile SBFspot.

**Comment by sbf-** [comment]
date is [UTC](https://www.epochconverter.com/) stored as int(4)
power in W
energy in Wh

**Comment by fritserasmus** [comment]
@bshaffer82 , once you have managed to sort this out and able to upload the usage to PVOUTPUT, please post the solution here so others can also benefit from that.

Your other option also is to update PVOUTPUT without logging the data to emonCMS (Or also to log the data)

**Comment by Bruut8** [comment]
Will this work when adding -finq to the daydata cron job? That should force the uploads to go on during the night right?

Edit: I see now this issue is over a year old. Hope someone finds time to reply.

**Comment by Bruut8** [comment]
To answer my own comment for those who need the info:

-finq does not force uploads during the night. 

The script stops updating data in the 'daydata' table when there is no more energy generated. Without data in that table, the 'vwpvodata' is not populated, hence the uploadscript has no data to upload.
To upload consumption data 24/7 you need to create a script to calculate the power consumption data and add that in the 'consumption' table every five minutes. As mentioned above, timestamp is in unixtime, power in W and energy in Wh. The energy value is cumulative (just like 'totalyield' in energy generation), pvoutput will automatically calculate the power generated between datapoints.
This data will then appear in the 'vwconsumption' view and (during daylight) in the 'vwpvodata'  view, where the 'nearest5min' is added. Alas, this is not happening when it is dark, because there is no data in 'daydata'.
The last step then is adding data in the 'daydata'  table manually when there is no output from the inverter. Make sure the timestamp is the same as the 'nearest5min' mentioned before, set the correct serialnumber and put 0 in energy- and powergeneration. Leave the pvoutput column at null (0 will not work).
The above step will ensure the 'vwpvodata' table is filled and uploaded. 

If there is an easier way please let me know, until then the above step will solve the problem without modifying the SBFspot code.

**Comment by aradovan** [comment]
> Yes, fill consumption table with Timestamp, Energy (Wh) and Power (W)
> The only challenge is the upload when it's dark and there is no production. Upload stops when production stops, so this is still an issue. This requires changes to the code.

Hi, firstly sorry about my english... I would like to get the current consumption as I do in sunny portal, but running "SBFspot -v -finq -nocsv" I'm unable find that value into the text result. I've also search for it in vwAvgConsumption table (also vwPvoData that joins it) but this table is empty. You tell about filling this table but my problem is how to get that consumption value to fill it or get it in text result... Is there some [-option] for showing it? I dont need to upload it, just read it each time I run the command...

**Reply by g-mocken** [reply]
I have the same question as @aradovan (or at least a related one), which unfortunately got no answer for two years. Let me elaborate a bit:

In a very similar way to the _internally_ measured production data, the Sunnyboy inverter also logs consumption data of an _externally_ connected EnergyMeter (or HomeManager). The data is shown in the same table in the webinterface and also pushed to SunnyPortal.

-> Is this logged consumption data accessible via the protocol used by SBFspot? Is SBFspot capable of retrieving it? If so, how can it be configured in such a way, and where does the data end up in the database? (Currently, consumption tables and views all are empty).
I am not interested in live data, but the logged consumption data. By consumption I mean the actually consumed total energy (imported and produced), not just the one im imported from the grid.

**Reply by sbf-** [reply]
> Is this logged consumption data accessible via the protocol used by SBFspot? Is SBFspot capable of retrieving it?

No. I'm not sure the consumption data is actually stored in the inverter.
Is this consumption data visible in SunnyExplorer?

**Reply by g-mocken** [reply]
No, there is no consumption log data in SunnyExplorer. I still believe that the data is stored in the inverter, because
- the data is present in the inverter's web interface
- day-data (5min bins) is available in the web interface for the exact same ~2 month period as for the production data, so it's very likely not an on-the-fly copy of sunny portal data (which has not such 2 month limit)
- there is no consumption data in the energy meter's web interface.

So my best guess is that the inverter continuously retrieves the energy meter's measurements, logs them along with the (internally measured) production data, and uploads both occasionally to sunny portal.

**Reply by sbf-** [reply]
Hmm, your conclusions sound valid...
Did some experiments on my inverter (no EM/HM involved) and it returns a list of empty data, so it looks like there is "some" table which is empty in my case.
If you're ready to test, let me know

**Reply by g-mocken** [reply]
I am very willing to test it, provided that the experimental version does not write anything back to the inverter nor otherwise messes with my regular setup, i.e. it can be installed in a different directory and use a separate database.

**Reply by g-mocken** [reply]
Any update on this?
